### PR TITLE
#2677 Introduce ClusterStatus to Collie

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2677)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala [hunk 1]
```diff
@@ -51,10 +51,6 @@ trait BrokerCollie extends Collie[BrokerClusterInfo] {
     implicit val exec: ExecutionContext = executionContext
     clusters().flatMap(clusters => {
       clusters
-        .filter(_._1.isInstanceOf[BrokerClusterInfo])
-        .map {
-          case (cluster, containers) => cluster.asInstanceOf[BrokerClusterInfo] -> containers
-        }
         .find(_._1.key == creation.key)
```
Comment (saivirtue) [inline ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala]: It seems we don't use the collections `clusters` anymore...
So we change this to `cluster(key)` ?

File: ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala [hunk 2]
```diff
@@ -142,21 +137,20 @@ trait StreamCollie extends Collie[StreamClusterInfo] {
...
-                        error = None,
-                        lastModified = CommonUtils.current()
+                      postCreate(
+                        new StreamClusterStatus(
+                          group = creation.group,
+                          name = creation.name,
+                          definition = definition,
+                          // no state means cluster is NOT running so we cleanup the dead nodes
+                          aliveNodes = state.map(_ => successfulContainers.map(_.nodeName).toSet).getOrElse(Set.empty),
+                          // We do not care the user parameters since it's stored in configurator already
```
Comment (saivirtue) [inline ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala]: I think this is a residue comment?

File: ohara-agent/src/main/scala/com/island/ohara/agent/ZookeeperCollie.scala [hunk 3]
```diff
@@ -126,19 +122,19 @@ trait ZookeeperCollie extends Collie[ZookeeperClusterInfo] {
...
-                      state = state,
-                      error = None,
-                      lastModified = CommonUtils.current()
+                    postCreate(
+                      new ZookeeperClusterStatus(
+                        group = creation.group,
+                        name = creation.name,
+                        // no state means cluster is NOT running so we cleanup the dead nodes
+                        aliveNodes = state.map(_ => successfulContainers.map(_.nodeName).toSet).getOrElse(Set.empty),
+                        // We do not care the user parameters since it's stored in configurator already
```
Comment (saivirtue) [inline ohara-agent/src/main/scala/com/island/ohara/agent/ZookeeperCollie.scala]: ditto

File: ohara-configurator/src/main/scala/com/island/ohara/configurator/route/ContainerRoute.scala [hunk 4]
```diff
@@ -43,10 +44,11 @@ object ContainerRoute {
...
                       clusterKey = ObjectKey.of(groupOption.getOrElse(GROUP_DEFAULT), clusterName),
                       clusterType = cluster match {
-                        case _: ZookeeperClusterInfo => "zookeeper"
-                        case _: BrokerClusterInfo    => "broker"
-                        case _: WorkerClusterInfo    => "worker"
-                        case _                       => "unknown"
+                        case _: ZookeeperClusterStatus => "zookeeper"
+                        case _: BrokerClusterStatus    => "broker"
+                        case _: WorkerClusterStatus    => "worker"
+                        case _: StreamClusterStatus    => "streamapp"
```
Comment (saivirtue) [inline ohara-configurator/src/main/scala/com/island/ohara/configurator/route/ContainerRoute.scala]: "streamapp" => "stream" :)

File: ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala (hunk 1 above)
Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala]: BrokerCollie and WorkerCollie have to handle both "adding" and "creating". The ```cluster(key)``` assumes that there is "already" a running cluster and it obstruct us from processing "creating"

File: ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala (hunk 2 above)
Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala]: you are right

File: ohara-configurator/src/main/scala/com/island/ohara/configurator/route/ContainerRoute.scala (hunk 4 above)
Comment (chia7712) [reply ohara-configurator/src/main/scala/com/island/ohara/configurator/route/ContainerRoute.scala]: roger that !

**Comment by chia7712** [comment]
retry k8sConfigurator

**Comment by chia7712** [comment]
retry otherIT

File: ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala (hunk 1 above)
Comment (saivirtue) [reply ohara-agent/src/main/scala/com/island/ohara/agent/BrokerCollie.scala]: > BrokerCollie and WorkerCollie have to handle both "adding" and "creating".

Got it :)

**Comment by chia7712** [comment]
The error is traced by #2852

**Pull request merged**